web/web.go

```diff
@@ -0,0 +1,771 @@
+		panic(fmt.Errorf("can't get database name"))
+	}
+	reportFileName := "frontend/" + filePrefix + "report.txt"
+	schemaFileName := "frontend/" + filePrefix + "schema.txt"
+	response := &schemaAndReportFile{}
+	conversion.WriteSchemaFile(app.conv, now, schemaFileName, ioHelper.Out)
+	conversion.Report(app.driver, nil, ioHelper.BytesRead, "", app.conv, reportFileName, ioHelper.Out)
+	reportAbsPath, err := filepath.Abs(reportFileName)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Can not create absolute path : %v", err), http.StatusInternalServerError)
+	}
+	schemaAbsPath, err := filepath.Abs(schemaFileName)
+	response.Report = reportAbsPath
+	response.Schema = schemaAbsPath
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(response)
+}
+
+type TableInterleaveStatus struct {
+	Possible bool
+	Parent   string
+	Comment  string
+}
+
+func checkPrimaryKeyPrefix(table string, refTable string, fk ddl.Foreignkey, tableInterleaveIssues *TableInterleaveStatus) bool {
+	childPks := app.conv.SpSchema[table].Pks
+	parentPks := app.conv.SpSchema[refTable].Pks
+	if len(childPks) >= len(parentPks) {
+		for i, pk := range parentPks {
+			if i >= len(fk.ReferColumns) || pk.Col != fk.ReferColumns[i] || pk.Col != childPks[i].Col || fk.Columns[i] != fk.ReferColumns[i] {
+				return false
+			}
+		}
+	} else {
+		return false
+	}
+	return true
+}
+func removeFk(slice []ddl.Foreignkey, s int) []ddl.Foreignkey {
+	return append(slice[:s], slice[s+1:]...)
+}
+
+func checkForInterleavedTables(w http.ResponseWriter, r *http.Request) {
+	table := r.FormValue("table")
+	if app.conv == nil || app.driver == "" {
+		http.Error(w, fmt.Sprintf("Schema is not converted or Driver is not configured properly. Please retry converting the database to spanner."), 404)
+		return
+	}
+	if table == "" {
+		http.Error(w, fmt.Sprintf("Table name is empty"), http.StatusBadRequest)
+	}
+	tableInterleaveIssues := &TableInterleaveStatus{Possible: true}
+	if _, found := app.conv.SyntheticPKeys[table]; found {
+		tableInterleaveIssues.Possible = false
+		tableInterleaveIssues.Comment = "Has synthetic pk"
+	}
+	if tableInterleaveIssues.Possible == true {
+		for i, fk := range app.conv.SpSchema[table].Fks {
+			refTable := fk.ReferTable
+			if _, found := app.conv.SyntheticPKeys[refTable]; found {
+				continue
+			}
+			ok := checkPrimaryKeyPrefix(table, refTable, fk, tableInterleaveIssues)
+			if ok == true {
+				tableInterleaveIssues.Parent = refTable
+				sp := app.conv.SpSchema[table]
+				sp.Parent = refTable
+				sp.Fks = removeFk(sp.Fks, i)
+				app.conv.SpSchema[table] = sp
+				break
+			}
+		}
+		if tableInterleaveIssues.Parent == "" {
+			tableInterleaveIssues.Possible = false
+			tableInterleaveIssues.Comment = "No valid prefix"
+		}
+	}
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(tableInterleaveIssues)
+}
+
+type App struct {
+	sourceDB *sql.DB
+	dbName   string
+	driver   string
+	conv     *internal.Conv
+}
+
+var app App
```

Comment:
Can you clarify the role of app -- looks like it is used to keep track of state from one request to the next.
I assume this means that this module is essentially single-session i.e. we can only have one stream of things going at one time because of this internal (and global) state?

Let's add a comment on app to describe its role and the single-session assumption it implies.

Reply:
Yes Nevin, it is used to maintain the state of the session and we can only have one session at a time, i'll add a comment to address this

Reply:
added comment on app to explain its role.

Reply:
Suggest slight refactoring of comment:
- put the field comments in the App struct (rather than with the var).
- slight reword of comment

Suggest:

```
type App struct {
	sourceDB    *sql.DB // Connection to source database in case of direct connection
	dbName      string // Name of source database
	driver      string // Name of HarbourBridge driver to use
	conv        *internal.Conv // Current conversion state
	sessionFile string // Path to session file
}

// app maintains the state of the session and there can be
// only one session at any given time.
var app App
```